raft-tech/TANF-app
# Issue 818 (PR #1): Implement ClamAV Client for use with uploads

### Summary of Changes
First part of a set of PRs to address #818 

Implements a ClamAV REST Client that can be used to send files from Django to the clamav-rest app instance / docker container.

This client enables several functions:
- Initializes a connection with the AV_SCAN_URL specified in Django settings
  - This URL may be overridden on initialization - which may prove to be useful for future testing asserting counter cases
- Creates a [requests.Session object](https://docs.python-requests.org/en/master/user/advanced/#session-objects) which is re-used between requests. This yields a significant performance increase because it can re-use the underlying TCP connection instead of creating a new one for each request to ClamAV
- Implements [urllib3.Retry](https://urllib3.readthedocs.io/en/latest/reference/urllib3.util.html) to allow retrying failed connections due to internal errors on ClamAV, connection issues, timeout errors, or rate limiting of requests. This hardens our implementation and helps to prevent temporary issues from causing dropped uploads.
- Provides a method `scan_file` which given a file or file-like object and its name will perform a request using the previously created session to POST the file to clamav-rest and return a boolean result indicating whether or not the file passed security checks (marked as CLEAN by Clam-AV)

In addition to this, in an effort to reduce flaky tests related to ClamAV the Docker-Compose `command` was updated to make a call to a new script `wait_for_services.sh` - this script uses the open source package [wait-for-it](https://pypi.org/project/wait-for-it/) to ensure that the services `clamav-rest` and `localstack` are accepting HTTP connections before the Docker container for the Django web server will be launched. This only impacts local and CI environments.  

### How to Test
* Pull this PR down and re-build docker containers
* Start up the web docker-compose service and ensure that the wait_for_services.sh script was run. Output should similar to this:
```
web_1          | [*] Waiting 15 seconds for clamav-rest:9000
web_1          | [+] clamav-rest:9000 is available after 0 seconds
web_1          | [*] Waiting 15 seconds for localstack:4566
web_1          | [+] localstack:4566 is available after 0 seconds
web_1          | ClamAV and Localstack are ready!
web_1          | Postgres is ready!
```
* Run the backend tests and confirm they all pass:
```
cd tdrs-backend
docker-compose -f docker-compose.yml -f docker-compose.local.yml run --rm web bash -c "pytest"
```

### Deliverable 1: Accepted Features

> *Performance Standard(s): At the beginning of each sprint, the Product Owner and development team will collaborate to define a set of user stories to be completed during the sprint. Acceptance criteria for each story will also be defined. The development team will deliver code and functionality to satisfy these user stories.*

> *Acceptable Quality Level: Delivered code meets the acceptance criteria for each user story. Incomplete stories will be assessed and considered for inclusion in the next sprint.*

 - [x] Files can be sent to ClamAV from Django code

As facilitator/product manager, @kniz-raft will decide if ACs are met from Raft's perspective.

### Deliverable 2: Tested Code

> *Performance Standard(s): Code delivered under the order must have substantial test code coverage. Version-controlled HHS GitHub repository of code that comprises products that will remain in the government domain.*

> *Acceptable Quality Level: Minimum of 90% test coverage of all code. All areas of code are meaningfully tested.*

+ Are all areas of code introduced in this PR meaningfully tested?
+ Are code coverage minimums met?

### Deliverable 3: Properly Styled Code

> *Performance Standard(s): GSA 18F Front- End Guide*

> *Acceptable Quality Level: 0 linting errors and 0 warnings*


### Deliverable 4: Accessible

> *Performance Standard(s): Web Content Accessibility Guidelines 2.1 AA standards*

> *Acceptable Quality Level: 0 errors reported using an automated scanner and 0 errors reported in manual testing*

N/A no UX changes

### Deliverable 5: Deployed

> *Performance Standard(s): Code must successfully build and deploy into the staging environment.*

> *Acceptable Quality Level: Successful build with a single command*

> NOTE: until we have a proper staging environment this may not be satisfiable prior to merging


### Deliverable 6: Documented

> *Performance Standard(s): Summary of user stories completed every two weeks. All dependencies are listed and the licenses are documented. Major functionality in the software/source code is documented, including system diagram. Individual methods are documented inline in a format that permits the use of tools such as JSDoc. All non-inherited 800-53 system security controls are documented in the Open Control or OSCAL format and HHS Section 508 Product Assessment Template (PAT) are updated as appropriate.*

> *Acceptable Quality Level: Combination of manual review and automated testing, if available*


### Deliverable 7: Secure

> *Performance Standard(s): Open Web Application Security Project (OWASP) Application Security Verification Standard 3.0*

> *Acceptable Quality Level: Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities*
